### DISCUSSION TITLE
Serverless (+ single route builds)

### DISCUSSION THREAD

**Discussion by thescientist13** [ProjectEvergreen/greenwood#626] [General]
## Overview
As an off shot from the discussion in https://github.com/ProjectEvergreen/greenwood/discussions/576 on the topic of SSR / SST, it would also be valuable to explore how to run Greenwood "serverlessly" in a Lambda or Netlify function.  This is super cool because it can help bridge the gap between the ends of the spectrum of SSG <> SSR.  Effectively, using something like [Netlify's DPR](https://github.com/jamstack/jamstack.org/discussions/549), you can build only parts of your site statically per build, and let more dynamic sites be built on demand and ached just as if it was all done at build time.

But of course you can fine tune this, but this is more up to the hosting provider.

## Thoughts
I think it comes down to a couple things from Greenwood side of it:
1. (primarily) Being able to build one page / route at a time, e.g.
    ```sh
    $ npx @greenwood/cli build --route /blog/yyyy/mm/dd/
    ```
 1. Being able to pre-render HTML on the fly _without_ needing to install Puppeteer in a serverless environment.  Both for issues of size / speed, but also due to environment / OS dependencies needed for Chromium.

**Comment by hutchgrant** [comment]
Modifying a component/style/template would also need to trigger builds on any/all routes where that component is present.

Cool idea.

**Reply by thescientist13** [reply]
For sure!

I think from my understanding of the the idea though is if you change templates / styles, that _would_ trigger a rebuild / redeploy.  It's just that say if you have 1000 blog posts, or even more extreme, a large publication with an archive of 100,000 articles, you can just build your "static pages" like Home, Contact, About, and everything that matches pattern `/blog/*` would just not get built.... _until_ it is actually user requested, then you would build that single page on the fly and cache it (stitching together that template one time for the content).

Which is pretty dope as now if content only changes for one route, you don't also _also_ have to build out every other blog post too.  💥 

If you change your app / templates, then yes, you would have to invalidate the caches and start the cycle over again.

**Comment by thescientist13** [comment]
Started making some progress here but having issues running the Lit shim on Netlify functions.  (not entirely unexpected I suppose)
https://github.com/thescientist13/lit-ssr-experiments/pull/5
https://deploy-preview-5--reverent-roentgen-ce17b4.netlify.app/.netlify/functions/lit-ssr

Not sure if it will help, but there are some discussions around [runtime agnostic WCs](https://github.com/webcomponents/community-protocols/issues/17), which may provide more developer ergonomics in the context of running serverless or maybe in #639 where NodeJS environments and globals may be more constrained.

**Comment by thescientist13** [comment]
Though I think I had some [luck already using Lambda with ESM recently](https://github.com/whatwg/dom/issues/831#issuecomment-983955120), this appears to confirm support for it officially.
https://aws.amazon.com/de/blogs/compute/using-node-js-es-modules-and-top-level-await-in-aws-lambda/

Which I assume by proxy means that Netlify would also get this support as well!  👀

**Comment by thescientist13** [comment]
Made some [good progress here](https://github.com/ProjectEvergreen/greenwood/pull/812), with a couple takeaways
1. Lambda dont have filesystems per se, so `fs.read` and `fs.write` functions have to be re-thought
1. This pops up more when introducing SSR and needing to generate the HTML on disk so we can re-reuse our Rollup implementation to generate the bundles from it - https://github.com/ProjectEvergreen/greenwood/pull/838/files#diff-d61c0039ae5955506325c361ce49501ca90c732a543cc2a45ac2b7de0ab93329R264

In otherwords, to go serverless, we'll really want to lean more into an "in-memory" architecture and avoid disk writing as much as possible, or somehow auto-detect this is a serverless context and just don't write out, like _graph.json_?

Additionally, for SSR, this means we can re-use built assets by tracking their build status in memory and avoid bundling entirely!

**Comment by thescientist13** [comment]
Additionally, we should incorporate support for streaming HTML with this technique, which may require a refactoring of our Koa server.  I think it will make sense to hold all this Serverless work for a 2.0 release?

**Comment by thescientist13** [comment]
We are landing Serverless support in an upcoming release (v0.29.0)!  You can check out the releases tab to try out the alpha releases and these associated issues to see what we've been up to:
- #1008 
- #1139 
- #1128
